runtime/parachains/src/dmp.rs

```diff
@@ -151,7 +180,8 @@ impl<T: Config> Pallet<T> {
 			return Err(QueueDownwardMessageError::ExceedsMaxMessageSize)
 		}
 
-		if DownwardMessageQueues::<T>::decode_len(para).unwrap_or(0) > MAX_MESSAGE_QUEUE_SIZE {
+		// Hard limit on Queue size
+		if Self::dmq_length(*para) > Self::dmq_max_length(config.max_downward_message_size) {
```

Comment:
This doesn't do what you think it does: `max_downward_message_size` is not about the queue size, but the _message_ size.